Feat: Operating a hApp (#595)
holochain/docs-pages: src/pages/build/operating-a-happ.md

* You can create an [Electron](https://www.electronjs.org/)-based binary for Windows, macOS, and Linux using the [Kangaroo repository](https://github.com/holochain/kangaroo-electron) as a starting point. <!-- TODO: link to guide -->
* You can create a [Tauri](https://tauri.app/)-based binary for Windows, macOS, Linux, and Android using the [p2p Shipyard](https://darksoil.studio/p2p-shipyard/) tool from our friends at [dark soil studio](https://darksoil.studio/). [Read their documentation](https://darksoil.studio/p2p-shipyard/guides/creating-an-app.html) to find out how.

Because a hApp is defined by its DNAs, and those DNAs live on the devices of your users, it's a good idea to use auto-updating because you can't easily update the hApp any other way. Both of the choices above support auto-updating, with two warnings:

!!! info Auto-updating Holochain or the hApp may create network forks

> * If you make changes to any of the integrity code in a DNA, it'll be backed by an entirely new network without access to the old network and its data. This means users who have upgraded may not be able to see users who haven't upgraded yet.

Author: What about updating a non-compatible Holochain version? I presume that works okay, with all the unintended consequences that go along with it? And would database migrations be handled too, as long as there's a migration path between versions?

The solution to this is **always-on nodes**. You can run a few hApp instances on devices you own (or in the cloud), or you can encourage some users to leave their devices turned on with their copy of the hApp running. <!-- TODO: Update this to mention full-arc when sharding is a thing -->

Our partners at Holo Hosting also offer [Cloud Nodes](https://holo.host/product/hosting-services/), which are always-on Holochain conductors hosted on a decentralized network of community-owned devices around the world. Click the 'Request a demo' button on [Holo's website](https://holo.host/) to get started.

## Running network infrastructure

### Bootstrap/signal server

For a hApp to run smoothly on the public internet, its users need a little bit of centralized infrastructure to discover each other and establish peer-to-peer connections. You'll need to set up and maintain an instance of the Kitsune2 bootstrap/signal server; [read the howto](#) <!-- TODO: link when https://github.com/holochain/docs-pages/pull/594 is merged --> to get set up.

> 

Author: Will be resolved after the mentioned PR is merged.

### WebRTC servers

> Although Holochain is designed to support multiple transport protocols, currently it only officially supports WebRTC over IP networks. We recommend you find or set up your own WebRTC [STUN](https://en.wikipedia.org/wiki/STUN) and [ICE](https://en.wikipedia.org/wiki/Interactive_Connectivity_Establishment) servers. There are a number of free public servers run by various companies; for instance, Holochain defaults to `stun.l.google.com` for both STUN and ICE if you don't specify any.

Reviewer: ```suggestion
Although Holochain is designed to support multiple transport protocols, currently it only officially supports WebRTC over IP networks. We recommend you find or set up your own WebRTC [STUN](https://en.wikipedia.org/wiki/STUN) and [ICE](https://en.wikipedia.org/wiki/Interactive_Connectivity_Establishment) servers. There are a number of free public servers run by various companies;
```

* You can create an [Electron](https://www.electronjs.org/)-based binary for Windows, macOS, and Linux using the [Kangaroo repository](https://github.com/holochain/kangaroo-electron) as a starting point. <!-- TODO: link to guide -->
* You can create a [Tauri](https://tauri.app/)-based binary for Windows, macOS, Linux, and Android using the [p2p Shipyard](https://darksoil.studio/p2p-shipyard/) tool from our friends at [dark soil studio](https://darksoil.studio/). [Read their documentation](https://darksoil.studio/p2p-shipyard/guides/creating-an-app.html) to find out how.

!!! info Be careful with auto-updating
The auto-updater code in Kangaroo checks your project's GitHub releases page for releases with a version number that's [semver](https://semver.org)-compatible with the one the user currently has installed. It's up to you to make sure that you bump your version numbers in a way that users expect.

> These things can cause issues with auto-updating:

Reviewer: ```suggestion
In particular, these things can cause issues with auto-updating:
```

> * Updated coordinator zomes in any DNA will not replace the existing coordinator zomes in-place, which may cause mismatches with updated UIs that tries to call those new zomes' functions. (Users without any installation history for the app will of course get the new coordinator zomes.)

Reviewer: ```suggestion
* Updating/changing the happ file that's bundled with the app will not replace the happ file for existing users of your app receiving an auto-upate. This means:
  - You cannot update coordinator zomes
  - If you change the happ file that you bundle with the app, any new users of your app that start from this new version will be using a different happ than existing users that installed the version via auto-updates.
```

Reviewer: And remove the point below. It seems more clear to me like this and the point below is not correct because participation in the old network will not be lost. There is simply no way to update integrity zomes for an existing user since the happ file won't be replaced through an auto-update.

Reviewer: Yes adding this line is great :+1:

> * Updating integrity zomes in any DNA will cause a new cell to be created, with a new DHT separate from the old one, causing the agent's old source chain and their participation in the old network to be lost.

Author: ```suggestion
```

> * Updating the bundled Holochain version to one with an incompatible database schema will cause the user's authoring history to be lost, unless that Holochain version has a migration path from the old schema to the new one.

Reviewer: ```suggestion
* Updating the bundled Holochain version to one with an incompatible database schema will cause the user's authoring history to be lost, unless that Holochain version has a migration path from the old schema to the new one.
```

> * Updating the bundled Holochain version to one with an incompatible database schema will cause the user's authoring history to be lost, unless that Holochain version has a migration path from the old schema to the new one.

Reviewer: ```suggestion
* You should only update the bundled holochain version to semver compatible holochain versions
```

* Updating the bundled Holochain version to one with an incompatible network protocol will prevent users on the new version of the app from communicating with users on the old version.

In all of these cases, we strongly recommend that you bump the leftmost integer of the app's version number, to indicate to the auto-updater that it's an incompatible version from the previous one.
!!!

## Keeping a DHT alive

A DHT only exists if agents are actively running cells instantiated from that DHT and communicating with each other.

> If you have enough users, your hApp's DHTs will probably stay alive by themselves. But if you have a small number of users, or if they're only online for a few minutes at a time, the network might suffer from poor connectivity, which your users will experience as inability to join the network, missing data from others, and network latency.

Author: https://github.com/holochain/docs-pages/pull/595/commits/cd7937a9176743d8537787a83447abe4e0642e04

* You can create an [Electron](https://www.electronjs.org/)-based binary for Windows, macOS, and Linux using the [Kangaroo repository](https://github.com/holochain/kangaroo-electron) as a starting point. <!-- TODO: link to guide -->
* You can create a [Tauri](https://tauri.app/)-based binary for Windows, macOS, Linux, and Android using the [p2p Shipyard](https://darksoil.studio/p2p-shipyard/) tool from our friends at [darksoil studio](https://darksoil.studio/). [Read their documentation](https://darksoil.studio/p2p-shipyard/guides/creating-an-app.html) to find out how.

!!! info Be careful with auto-updating
The auto-updater code in Kangaroo checks your project's GitHub releases page for releases with a version number that's [semver](https://semver.org)-compatible with the one the user currently has installed. It's up to you to make sure that you bump your version numbers in a way that an auto-update doesn't break anything.

In particular, the following needs to be considered:

> * Updating/changing the happ file that's bundled with the app will not replace the hApp file for existing users of your app receiving an auto-update. This means:

Author: ```suggestion
* Updating/changing the hApp file that's bundled with the app will not replace the hApp file for existing users of your app receiving an auto-update. This means:
```
